dev/src/bulk-writer.ts

```diff
@@ -211,25 +221,70 @@ class BulkCommitBatch {
     // Capture the error stack to preserve stack tracing across async calls.
     const stack = Error().stack!;
 
-    return this.writeBatch.bulkCommit().catch(err => {
-      throw wrapError(err, stack);
-    });
+    let results: BatchWriteResult[] = [];
+    for (let attempt = 0; attempt < MAX_RETRY_ATTEMPTS; attempt++) {
+      await this.backoff.backoffAndWait();
+
+      try {
+        results = await this.writeBatch.bulkCommit();
+      } catch (err) {
+        // Map the failure to each individual write's result.
+        results = [...this.pendingOps.keys()].map(key => {
+          return {key, writeTime: null, status: wrapError(err, stack)};
+        });
+      }
+      this.processResults(results, /* shouldRetry= */ true);
+
+      if (this.pendingOps.size > 0) {
+        logger(
+          'BulkWriter.bulkCommit',
+          null,
+          `Current batch failed at retry #${attempt}. Num failures: ` +
+            `${this.pendingOps.size}.`
+        );
+
+        this.writeBatch = new WriteBatch(this.firestore, this.writeBatch, [
+          ...this.pendingOps.keys(),
+        ]);
+      } else {
+        this.completedDeferred.resolve();
+        return;
+      }
+    }
+
+    this.processResults(results, /* shouldRetry= */ false);
+    this.completedDeferred.resolve();
   }
 
   /**
    * Resolves the individual operations in the batch with the results.
    */
-  processResults(results: BatchWriteResult[], error?: Error): void {
-    if (error === undefined) {
-      for (let i = 0; i < this.opCount; i++) {
-        this.resultsMap.get(i)!.resolve(results[i]);
-      }
-    } else {
-      for (let i = 0; i < this.opCount; i++) {
-        this.resultsMap.get(i)!.reject(error);
+  processResults(results: BatchWriteResult[], shouldRetry: boolean): void {
```

Comment:
s/shouldRetry/mayRetry (or allowRetry)?

Also might be worth adding `@param` comments.

Reply:
N/A since we're using `failRemainingOperations`.